File: src/com/train/algorithm/twoPointers/BoatstoSavePeople.cpp
//
// Created by 罗斯 on 28/8/2022.
//
#include <vector>

using namespace std;

class Solution {
public:
    int numRescueBoats(vector<int>& people, int limit) {
//        sort(people.rbegin(),people.rend());
//        int i = 0;
//        int j = people.size() - 1;
//        int res = 0;
//        while(i <= j){
//            ++res;
//            if(i == j) break;
//            if(people[i++] + people[j] <= limit) --j;
//        }
//        return res;


        sort(people.begin(), people.end());
        int i = 0;
        int j = people.size() - 1;
        int res = 0;
        while(i <= j){
            if(people[i] + people[j] <= limit){
                ++i;
                --j;
            }else
                --j;

            ++res;
        }
        return res;
    }
};

File: src/com/train/algorithm/twoPointers/implementInPython/BoatsToSavePeople.py
class Solution(object):
    def numRescueBoats(self, people, limit):
        people.sort()
        l,r,res = 0, len(people) - 1,0
        while l <= r:
            if people[l] + people[r] <= limit:
                l += 1
                r -= 1
            else:
                r -= 1
            res += 1
        return res
